pkg/controller/mizar/mizar-service-controller.go

```diff
@@ -0,0 +1,292 @@
+
+	switch event {
+	case EventType_Create:
+		err = c.processServiceCreation(svc, key)
+	case EventType_Update:
+		err = c.processServiceUpdate(svc, key)
+	case EventType_Delete:
+		err = c.processServiceDeletion(key)
+	default:
+		utilruntime.HandleError(fmt.Errorf("Unable to process service %v %v", event, key))
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *MizarServiceController) processServiceCreation(service *v1.Service, key string) error {
+	netName := strings.Split(service.Name, dnsServiceDefaultName+"-")[0]
+	fmt.Println("processServiceDeletion network name is %v", netName)
+
+	msg := &BuiltinsServiceMessage{
+		Name:          service.Name,
+		ArktosNetwork: netName,
+		Namespace:     service.Namespace,
+		Tenant:        service.Tenant,
+		Ip:            "",
+	}
+
+	response := GrpcCreateService(c.grpcHost, msg)
+	code := response.Code
+	//To Add: context := response.Message and then convert to json to get ip address
+	if code != CodeType_OK {
+		return errors.New("Service creation failed on Mizar side")
+	}
+
+	svcName := "context.name"
+	namespace := "context.namespace"
+	tenant := "context.tenant"
+	ip := "context.ip"
+
+	svc, err := c.serviceLister.ServicesWithMultiTenancy(namespace, tenant).Get(svcName)
+	if err != nil {
+		return err
+	}
+
+	if _, hasDNSServiceLabel := svc.Labels[labelDNSService]; hasDNSServiceLabel {
+		net, err := c.netLister.NetworksWithMultiTenancy(tenant).Get(netName)
+		if err != nil {
+			return err
+		}
+
+		if len(net.Status.DNSServiceIP) == 0 {
+			netReady := net.DeepCopy()
+			netReady.Status.DNSServiceIP = ip
+			_, err = c.netClientset.ArktosV1().NetworksWithMultiTenancy(tenant).UpdateStatus(netReady)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	svcToUpdate := svc.DeepCopy()
+	svcToUpdate.Spec.ClusterIP = ip
+	svcUpdated, err := c.kubeClientset.CoreV1().ServicesWithMultiTenancy(namespace, tenant).Update(svcToUpdate)
+	if err != nil {
+		klog.Infof("The following service failed to update: %v", svcUpdated)
+		return err
+	}
+
+	return nil
+}
+
+func (c *MizarServiceController) processServiceUpdate(service *v1.Service, key string) error {
+	netName := strings.Split(service.Name, dnsServiceDefaultName+"-")[0]
+	fmt.Println("processServiceDeletion network name is %v", netName)
+
+	msg := &BuiltinsServiceMessage{
+		Name:          service.Name,
+		ArktosNetwork: netName,
+		Namespace:     service.Namespace,
+		Tenant:        service.Tenant,
+		Ip:            "",
+	}
+	response := GrpcUpdateService(c.grpcHost, msg)
+	code := response.Code
+	//To Add: context := response.Message and then convert to json to get ip address
+	if code != CodeType_OK {
+		return errors.New("Service update failed on Mizar side")
+	}
+
+	return nil
+}
+
+func (c *MizarServiceController) processServiceDeletion(key string) error {
+	tenant, namespace, name, err := cache.SplitMetaTenantNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	netName := strings.Split(name, dnsServiceDefaultName+"-")[0]
```

Comment:
Many mizar controllers need to extract the netName from the service name. The way we are doing it is really hacky.

 It would make sense to create a util func for this purpose in util.go.  Some error handling should also be done in that util func, like if the netName does not start with "kube-dns", or the netName turns out to an empty string.